src/ng/messages/privmsg.rs

```diff
@@ -0,0 +1,251 @@
+use crate::{
+    color::Color,
+    ng::{FromIrcMessage, InvalidMessage, IrcMessage, Str, StrIndex, TagIndices, Tags, Validator},
+    parse_badges, parse_badges_iter, parse_emotes, Badge, BadgeInfo, BadgeKind, Emotes,
+};
+
+#[derive(Debug, Clone, PartialEq)]
+#[cfg_attr(feature = "serde", derive(::serde::Serialize, ::serde::Deserialize))]
+pub enum Ctcp<'a> {
+    Action,
+    Unknown { command: &'a str },
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Privmsg<'t> {
+    raw: Str<'t>,
+    tags: TagIndices,
+    name: StrIndex,
+    channel: StrIndex,
+    data: StrIndex,
+    ctcp: Option<StrIndex>,
+}
+
+impl<'t> Privmsg<'t> {
+    raw!();
+    tags!();
+    str_field!(name);
+    str_field!(channel);
+    str_field!(data);
+
+    pub fn ctcp(&self) -> Option<Ctcp<'_>> {
+        const ACTION: &str = "ACTION";
+        let command = &self.raw[self.ctcp?];
+        if command == ACTION {
+            Some(Ctcp::Action)
+        } else {
+            Some(Ctcp::Unknown { command })
+        }
+    }
+
+    pub fn is_action(&self) -> bool {
+        match self.ctcp() {
+            Some(Ctcp::Action) => true,
+            _ => false,
+        }
+    }
+
+    pub fn badge_info(&'t self) -> Vec<BadgeInfo<'t>> {
+        self.tags()
+            .get("badge-info")
+            .map(parse_badges)
+            .unwrap_or_default()
+    }
+
+    pub fn badges(&'t self) -> Vec<Badge<'t>> {
+        self.tags()
+            .get("badges")
+            .map(parse_badges)
+            .unwrap_or_default()
+    }
+
+    pub fn bits(&self) -> Option<u64> {
+        self.tags().get_parsed("bits")
+    }
+
+    pub fn color(&self) -> Option<Color> {
+        self.tags().get_parsed("color")
+    }
+
+    pub fn display_name(&'t self) -> Option<&str> {
+        self.tags().get("display-name")
+    }
+
+    pub fn emotes(&self) -> Vec<Emotes> {
+        self.tags()
+            .get("emotes")
+            .map(parse_emotes)
+            .unwrap_or_default()
+    }
+
+    pub fn is_broadcaster(&self) -> bool {
+        self.contains_badge(BadgeKind::Broadcaster)
+    }
+
+    pub fn is_moderator(&self) -> bool {
+        self.tags().get_as_bool("mod")
```

Comment:
should be
```suggestion
        self.contains_badge(BadgeKind::Moderator)
```

which should be equivalent according to twitch.